File: rust/src/leet_code/200.rs
//! This module provides a solution for the problem of counting the number of islands in a 2D grid.
//! Each cell in the grid can be '1' (land) or '0' (water). An island is surrounded by water and is
//! formed by connecting adjacent lands horizontally or vertically. This module provides a function
//! `num_islands` that counts the number of islands.

/// Struct `Solution` is a placeholder for the solution methods.
impl Solution {
    /// Counts the number of islands in a 2D grid map.
    ///
    /// # Parameters
    ///
    /// * `grid` - A 2D vector of characters. Each character can be '1' (land) or '0' (water).
    ///
    /// # Returns
    ///
    /// * An `i32` integer representing the number of islands in the grid.
    ///
    /// # Example
    ///
    /// ```
    /// let grid = vec![
    ///     vec!['1', '1', '1', '1', '0'],
    ///     vec!['1', '1', '0', '1', '0'],
    ///     vec!['1', '1', '0', '0', '0'],
    ///     vec!['0', '0', '0', '0', '0'],
    /// ];
    /// assert_eq!(Solution::num_islands(grid), 1);
    /// ```
    pub fn num_islands(mut grid: Vec<Vec<char>>) -> i32 {
        let (len_outer, len_inner) = (grid.len(), grid[0].len());
        let mut _grid = grid
            .iter()
            .enumerate()
            .flat_map(|(id, e)| e.iter().enumerate().map(move |(id2, val)| ((id, id2), val)))
            .filter(|(_, val)| *val == &'1');
        let mut output = 0;
        let mut cache = Vec::<(usize, usize)>::with_capacity(len_inner * len_outer);
        let mut _cache = vec![false; len_inner * len_outer];
        while let Some(((i, j), _)) = _grid.next() {
            if !_cache[i * len_inner + j] {
                output += 1;
                cache.push((i, j));
            }
            while let Some((i, j)) = cache.pop() {
                _cache[i * len_inner + j] = true;
                if i + 1 < len_outer && !_cache[(i + 1) * len_inner + j] && grid[i + 1][j] == '1' {
                    cache.push((i + 1, j));
                }
                if i > 0 && !_cache[(i - 1) * len_inner + j] && grid[i - 1][j] == '1' {
                    cache.push((i - 1, j));
                }
                if j + 1 < len_inner && !_cache[i * len_inner + j + 1] && grid[i][j + 1] == '1' {
                    cache.push((i, j + 1));
                }
                if j > 0 && !_cache[i * len_inner + j - 1] && grid[i][j - 1] == '1' {
                    cache.push((i, j - 1));
                }
            }
        }
        output
    }
}

File: java/leet_code/200.java
import java.util.LinkedList;
import java.util.Queue;

/**
 * Solution class for LeetCode problem 200: Number of Islands.
 *
 * <p>This class provides a method to count the number of islands in a 2D grid map. An island is
 * surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You
 * may assume all four edges of the grid are all surrounded by water.
 *
 * <p>Example usage:
 * <pre>
 *   Solution solution = new Solution();
 *   char[][] grid = {{'1','1','1','1','0'}, {'1','1','0','1','0'}, {'1','1','0','0','0'}, {'0','0','0','0','0'}};
 *   int numberOfIslands = solution.numIslands(grid);
 *   System.out.println(numberOfIslands);  // Outputs: 1
 * </pre>
 */
class Solution {

    /**
     * Counts the number of islands in the given 2D grid map.
     *
     * @param grid the 2D grid map of '1's (land) and '0's (water)
     *
     * @return the number of islands
     *
     * @throws NullPointerException if grid is null
     */
    public int numIslands(final char[][] grid) {
        if (grid == null || grid.length == 0) {
            throw new NullPointerException("Grid cannot be null or empty");
        }

        final int rows = grid.length;
        final int cols = grid[0].length;
        int numberOfIslands = 0;

        // Iterate over the entire grid
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                // If the current cell is land ('1'), start a new island
                if ((int) '1' == (int) grid[row][col]) {
                    numberOfIslands++;
                    grid[row][col] = '0'; // mark the current cell as visited
                    final Queue<int[]> neighbours = new LinkedList<>();
                    neighbours.add(new int[]{row, col});

                    // Perform BFS to visit all cells in the current island
                    while (!neighbours.isEmpty()) {
                        final int[] id = neighbours.remove();
                        final int r = id[0];
                        final int c = id[1];

                        // Check the four directions around the current cell
                        if (0 <= r - 1 && '1' == (int) grid[r - 1][c]) {
                            neighbours.add(new int[]{r - 1, c});
                            grid[r - 1][c] = '0';
                        }
                        if (r + 1 < rows && (int) '1' == (int) grid[r + 1][c]) {
                            neighbours.add(new int[]{r + 1, c});
                            grid[r + 1][c] = '0';
                        }
                        if (0 <= c - 1 && (int) '1' == (int) grid[r][c - 1]) {
                            neighbours.add(new int[]{r, c - 1});
                            grid[r][c - 1] = '0';
                        }
                        if (c + 1 < cols && (int) '1' == (int) grid[r][c + 1]) {
                            neighbours.add(new int[]{r, c + 1});
                            grid[r][c + 1] = '0';
                        }
                    }
                }
            }
        }

        return numberOfIslands;
    }
}
